components/greaselion/browser/greaselion_download_service.cc

```diff
@@ -0,0 +1,209 @@
+      ParsePrecondition(preconditions_value, kTwitterTips);
+  // Can't use URLPatternSet::Populate here because it does not expose
+  // any way to set the ParseOptions, which we need to do to support
+  // eTLD wildcarding.
+  for (const auto& urls_it : urls_value->GetList()) {
+    URLPattern pattern;
+    pattern.SetValidSchemes(URLPattern::SCHEME_HTTP | URLPattern::SCHEME_HTTPS);
+    if (pattern.Parse(urls_it.GetString(),
+                      URLPattern::ALLOW_WILDCARD_FOR_EFFECTIVE_TLD) !=
+        URLPattern::ParseResult::kSuccess) {
+      LOG(ERROR) << "Malformed pattern in Greaselion configuration";
+      urls_.ClearPatterns();
+      return;
+    }
+    urls_.AddPattern(pattern);
+  }
+  scoped_refptr<base::SequencedTaskRunner> task_runner =
+      base::CreateSequencedTaskRunnerWithTraits(
+          {base::MayBlock(), base::TaskPriority::USER_VISIBLE,
+           base::TaskShutdownBehavior::SKIP_ON_SHUTDOWN});
+  for (const auto& scripts_it : scripts_value->GetList()) {
+    base::FilePath script_path =
+        root_dir.AppendASCII(kGreaselionConfigFileVersion)
+            .AppendASCII(scripts_it.GetString());
+    if (script_path.ReferencesParent()) {
+      LOG(ERROR) << "Malformed filename in Greaselion configuration";
+    } else {
+      // Read script file on task runner to avoid file I/O on main thread.
+      auto script_contents = std::make_unique<std::string>();
+      std::string* buffer = script_contents.get();
+      base::PostTaskAndReplyWithResult(
+          task_runner.get(), FROM_HERE,
+          base::BindOnce(&base::ReadFileToString, script_path, buffer),
+          base::BindOnce(&GreaselionRule::AddScriptAfterLoad,
+                         weak_factory_.GetWeakPtr(),
+                         std::move(script_contents)));
+    }
+  }
+}
+
+GreaselionRule::~GreaselionRule() = default;
+
+void GreaselionRule::AddScriptAfterLoad(std::unique_ptr<std::string> contents,
+                                        bool did_load) {
+  if (!did_load || !contents) {
+    LOG(ERROR) << "Could not load Greaselion script";
+    return;
+  }
+  scripts_.push_back(*contents);
+}
+
+bool GreaselionRule::PreconditionFulfilled(
+    GreaselionPreconditionValue precondition,
+    bool value) const {
+  switch (precondition) {
+    case kMustBeTrue:
+      return value;
+    case kMustBeFalse:
+      return !value;
+    default:
+      return true;
+  }
+}
+
+bool GreaselionRule::Matches(const GURL& url, GreaselionFeatures state) const {
+  if (!PreconditionFulfilled(preconditions_.rewards_enabled,
+                             state[greaselion::REWARDS]))
+    return false;
+  if (!PreconditionFulfilled(preconditions_.twitter_tips_enabled,
+                             state[greaselion::TWITTER_TIPS]))
+    return false;
+  return urls_.MatchesURL(url);
+}
+
+void GreaselionRule::Populate(std::vector<std::string>* scripts) const {
+  scripts->insert(scripts->end(), scripts_.begin(), scripts_.end());
```

Comment:
can we avoid making a copy of the script here? Maybe keep an index from `scripts_`? Just concerned about limit memory on mobile cc @SergeyZhukovsky